pkg/trace/info/stats.go

```diff
@@ -155,15 +167,108 @@ func (ts *TagStats) publish() {
+		keys[i] = k
+		i++
+	}
+	sort.Strings(keys)
+
+	var results []string
+	for _, key := range keys {
+		value := statsMap[key]
+		if value > 0 {
+			results = append(results, fmt.Sprintf("%s:%d", key, value))
+		}
+	}
+
+	return strings.Join(results, ", ")
+}
+
+// TracesDropped contains counts for reasons traces have been dropped
+type TracesDropped struct {
+	DecodingError int64
+	EmptyTrace    int64
+	TraceIDZero   int64
+	SpanIDZero    int64
+	ForeignSpan   int64
+}
+
+func (s *TracesDropped) tagValues() (result map[string]int64) {
+	return map[string]int64{
+		"decoding_error": atomic.LoadInt64(&s.DecodingError),
+		"empty_trace":    atomic.LoadInt64(&s.EmptyTrace),
+		"trace_id_zero":  atomic.LoadInt64(&s.TraceIDZero),
+		"span_id_zero":   atomic.LoadInt64(&s.SpanIDZero),
+		"foreign_span":   atomic.LoadInt64(&s.ForeignSpan),
+	}
+}
+
+func (s *TracesDropped) String() string {
+	return inlineNonZeroTagValues(s.tagValues())
+}
+
+// TracesMalformed contains counts for reasons malformed traces have been accepted after applying automatic fixes
+type TracesMalformed struct {
+	DuplicateSpanID       int64
+	ServiceEmpty          int64
+	ServiceTruncate       int64
+	ServiceInvalid        int64
+	SpanNameEmpty         int64
+	SpanNameTruncate      int64
+	SpanNameInvalid       int64
+	ResourceEmpty         int64
+	TypeTruncate          int64
+	InvalidStartDate      int64
+	InvalidDuration       int64
+	InvalidHTTPStatusCode int64
+}
+
+// tagValues converts TracesMalformed into a map
+func (s *TracesMalformed) tagValues() (result map[string]int64) {
+	return map[string]int64{
+		"duplicate_span_id":        atomic.LoadInt64(&s.DuplicateSpanID),
+		"service_empty":            atomic.LoadInt64(&s.ServiceEmpty),
+		"service_truncate":         atomic.LoadInt64(&s.ServiceTruncate),
+		"service_invalid":          atomic.LoadInt64(&s.ServiceInvalid),
+		"span_name_empty":          atomic.LoadInt64(&s.SpanNameEmpty),
+		"span_name_truncate":       atomic.LoadInt64(&s.SpanNameTruncate),
+		"span_name_invalid":        atomic.LoadInt64(&s.SpanNameInvalid),
+		"resource_empty":           atomic.LoadInt64(&s.ResourceEmpty),
+		"type_truncate":            atomic.LoadInt64(&s.TypeTruncate),
+		"invalid_start_date":       atomic.LoadInt64(&s.InvalidStartDate),
+		"invalid_duration":         atomic.LoadInt64(&s.InvalidDuration),
+		"invalid_http_status_code": atomic.LoadInt64(&s.InvalidHTTPStatusCode),
+	}
+}
+
+func (s *TracesMalformed) String() string {
+	return inlineNonZeroTagValues(s.tagValues())
 }
 
 // Stats holds the metrics that will be reported every 10s by the agent.
 // Its fields require to be accessed in an atomic way.
 type Stats struct {
 	// TracesReceived is the total number of traces received, including the dropped ones.
 	TracesReceived int64
-	// TracesDropped is the number of traces dropped.
-	TracesDropped int64
+	// TracesDropped contains stats about the count of dropped traces by reason
+	TracesDropped TracesDropped
```

Comment:
Let's make them (`Traces{Dropped,Malformed}`) pointers, so we don't create copies every time we reference them.

Reply:
Are you sure it creates copies? If `ts.TracesDropped` created a copy every time you accessed it, then `atomic.AddInt64(&ts.TracesDropped.TraceIDZero, 1)` shouldn't work because it would be incrementing a copy?

Reply:
That’s because ts is a pointer, and adding the & sign takes the address of the field. The copy happens on assignment. It’s better to copy a pointer than a struct. Not a problem now but I think better be safe?

Reply:
Yeah makes sense. Updated to use pointers.